### PR TITLE
Fixed compatibility issue with windows 7

### PR DESCRIPTION
**Pull request by kampeador**
https://github.com/libuv/libuv/issues/3260

### CODE REVIEW

File: src/win/util.c [hunk 1]
```diff
@@ -1665,17 +1665,16 @@ int uv_os_unsetenv(const char* name) {
...
   int convert_result;
 
   if (buffer == NULL || size == NULL || *size == 0)
     return UV_EINVAL;
 
-  uv__once_init(); /* Initialize winsock */
-
-  if (GetHostNameW(buf, UV_MAXHOSTNAMESIZE) != 0)
-    return uv_translate_sys_error(WSAGetLastError());
+  if (GetComputerNameExW(ComputerNameDnsFullyQualified, buf, &buf_len) == 0)
```
Comment (vtjnash) [inline src/win/util.c:1676]: Do we know if this should be an FQDN, or perhaps the truncated NetBIOS name, or just the DNS hostname?
 ```suggestion
  if (GetComputerNameExW(ComputerNameNetBIOS, buf, &buf_len) == 0)
```

Comment (kampeador) [reply src/win/util.c:1676]: ComputerNameDnsHostname can be a good alternative too. There is a more detailed explanation in the separate post.

**Comment by kampeador** [comment]
Thanks for your review.
I'll explain.
I've made a typo here. ComputerNameDnsFullyQualified is gethostname+getdomainname on linux.

ComputerNameNetBIOS is probably the closest variant to gethostname, but incomplete.
ComputerNameDnsHostname will be correct most of the time. This is what many developers were using before GetAddrInfoW existed.
There are some differences:
1. Both GetHostName and GetComputerNameEx results can be overwritten by environment variable when working with clusters.
GetComputerNameEx with ComputerNameNetBIOS and GetHostName work the same here.
ComputerNameDnsHostname and others also use environment variable, but with different name: cluster hostname instead of just cluster name.
2. ComputerNameNetBIOS returnes truncated output (15), while GetHostName is limited by 255.
ComputerNameDnsHostname is capped by value inside registry.
If hostname length value is higher than 255, then GetHostName will fallback to NetBIOS name.
To handle this case more reliably we need to use WSALookupServiceBeginW api. Thats why GetHostName depends on winsock.
3. GetComputerNameEx and GetHostName use different registry values, but they get synchronized after restart. It is possible to "break" them, by editing registry and thus output will be different.

To emulate GetHostName with corner cases we need to:
1. Check environment variable for clusters.
2. If it fails, then use WSALookupServiceBeginW
3. If it fails, fallback to GetComputerNameEx(ComputerNameNetBIOS).

This will(probably) not get accepted anyway. Due to the fact that Windows is a closed source OS, 100% emulation is not trivial to achieve even with old leaked sources. There were small changes here and there between versions.
This approach is more reliable than original asci gethostname, so I'll leave it as reference.

**Comment by kampeador** [comment]
I've took a different approach and implemented a complete drop-in replacement for this function.
It turned out it is just a POSIX-like abstraction level implemented on top of public APIs available for a long time.
This is basically the same level as libuv, but closed source and requires Win 8.1+.

**Review by vtjnash — COMMENTED** [review]
CI: https://ci.nodejs.org/job/libuv-test-commit-windows-cmake/395/

some style nits to fix:

- The `*` leans left as in `WSAQUERYSETW* pwsaq`
- Tight formatting for if/else brackets:
```
if (...) {
} else {
}
```
- No parens for `return`
- single empty newlines inside functions (double empty newlines go between functions)
- local_buf_len initialization needs to be a constant (not constexpr), or moved to the function body. Same for `pwsaq`.
- Comments in C (`/**/`) style, not C++ (`//`)

File: src/win/util.c [hunk 2]
```diff
@@ -1663,6 +1664,138 @@ int uv_os_unsetenv(const char* name) {
...
+  int result_len;
+  int error_code = NO_ERROR;
+
+  // WSALookupService stuff
+  GUID guid_host_name = SVCID_HOSTNAME;
+  AFPROTOCOLS af_protocols[2] = { {AF_INET, IPPROTO_UDP},
+                                  {AF_INET, IPPROTO_TCP} };
+  // Avoid dynamic memory allocation if possible
+  const size_t local_buf_len = sizeof(WSAQUERYSETW) + 512;
+  CHAR local_buf[local_buf_len];
```
Comment (vtjnash) [inline src/win/util.c:1678]: VLA are only part of C99 (not before or after)

File: src/win/util.c [hunk 3]
```diff
@@ -1663,6 +1664,138 @@ int uv_os_unsetenv(const char* name) {
 }
 
 
+int uv__gethostnamew_nt60(WCHAR* name, int name_len)
```
Comment (vtjnash) [inline src/win/util.c:1667]:
```suggestion
static int uv__gethostnamew_nt60(WCHAR* name, int name_len)
```

File: src/win/util.c [hunk 4]
```diff
@@ -1674,7 +1807,7 @@ int uv_os_gethostname(char* buffer, size_t* size) {
 
   uv__once_init(); /* Initialize winsock */
 
-  if (GetHostNameW(buf, UV_MAXHOSTNAMESIZE) != 0)
+  if (uv__gethostnamew_nt60(buf, UV_MAXHOSTNAMESIZE) != 0)
```
Comment (vtjnash) [inline src/win/util.c:1810]: @libuv/windows Is it worthwhile to use the Windows 8 function, when available, and only polyfill for <= 7:

```suggestion
  sGetHostNameW gethostnamew = pGetHostNameW == NULL ? uv__gethostnamew_nt60 : pGetHostNameW;
  if (gethostnamew(buf, UV_MAXHOSTNAMESIZE) != 0)
```

And in winapi.c
```
  ws2_32_module = LoadLibraryA("ws2_32.dll");
  if (ws2_32_module != NULL) {
    pGetHostNameW = (sGetHostNameW)
      GetProcAddress(ws2_32_module, "GetHostNameW");
```

Comment (kampeador) [reply src/win/util.c:1808]: uv__gethostnamew_nt60 is basically a reverse engineered GetHostNameW to remove Win 8.1+ requirement.
They should behave exactly the same on Microsoft platforms.
GetProcAddress(ws2_32_module, "GetHostNameW"); will add an unnecessary branch in this case.
All 3 stages are also documented here:
https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-gethostname

However there can be some issues with alternative win32 implementations like wine.
GetHostNameW is just a wrapper for unix gethostname. It doesn't go through those stages defined in the documentation.
WSALookupService functions(stage 2) are not implemented, so my implementation will fallback to stage 3.

Comment (vtjnash) [reply src/win/util.c:1808]: Yes, wine is a reasonable case for it, as well as adopting any changes if Microsoft decides to change it

File: src/win/util.c [hunk 5]
```diff
@@ -1663,6 +1673,127 @@ int uv_os_unsetenv(const char* name) {
...
 
 
+static int uv__gethostnamew_nt60(WCHAR* name, int name_len) {
+  int result_len;
+  int error_code = NO_ERROR;
+
+  /* WSALookupService stuff
+   * Avoid dynamic memory allocation if possible */
+  CHAR local_buf[WSAQ_LOCAL_BUF_LEN];
+  DWORD dwlen = (DWORD)WSAQ_LOCAL_BUF_LEN;
```
Comment (vtjnash) [inline src/win/util.c:1683]: These are still illegal syntax expressions (for C89)

File: src/win/util.c [hunk 6]
```diff
@@ -1663,6 +1673,127 @@ int uv_os_unsetenv(const char* name) {
...
+                                       name_len);
+  if (result_len != 0) {
+    if (result_len > name_len) {
+      error_code = WSAEFAULT;
+    }
+    goto cleanup;
+  }
+
+  /* Stage 2: Do normal lookup through WSALookupServiceLookup */
+  pwsaq = (WSAQUERYSETW*)local_buf;
```
Comment (vtjnash) [inline src/win/util.c:1716]: style nit: space before cast
```suggestion
  pwsaq = (WSAQUERYSETW*) local_buf;
```

File: src/win/util.c [hunk 7]
```diff
@@ -1663,6 +1673,127 @@ int uv_os_unsetenv(const char* name) {
...
+      result_name = pwsaq->lpszServiceInstanceName;
+    } else {
+      error_code = WSAGetLastError();
+
+      if ((error_code == WSAEFAULT) && (dwlen > WSAQ_LOCAL_BUF_LEN)) {
+        /* Should never happen */
+        assert(sizeof(CHAR) * dwlen >= sizeof(WSAQUERYSETW));
+
+        /* Fallback to the heap allocation */
+        heap_data = (WSAQUERYSETW*)uv__malloc(sizeof(CHAR) * (size_t)dwlen);
```
Comment (vtjnash) [inline src/win/util.c:1739]: style nit: no unnecessary casts
```suggestion
        heap_data = uv__malloc(sizeof(CHAR) * (size_t) dwlen);
```

File: src/win/util.c [hunk 8]
```diff
@@ -1663,6 +1673,127 @@ int uv_os_unsetenv(const char* name) {
...
+    WSALookupServiceEnd(hlookup);
+
+    if(error_code != NO_ERROR) {
+      WSASetLastError(error_code);
+    }
+  }
+
+  if (result_name != NULL)
+  {
+    size_t wlen = wcslen(result_name) + 1;
```
Comment (vtjnash) [inline src/win/util.c:1763]: style nit
```suggestion
  if (result_name != NULL) {
    size_t wlen = wcslen(result_name) + 1;
```

File: src/win/util.c [hunk 9]
```diff
@@ -1663,6 +1673,127 @@ int uv_os_unsetenv(const char* name) {
...
+          }
+        } else {
+          error_code = WSA_NOT_ENOUGH_MEMORY;
+        }
+      }
+    }
+
+    WSALookupServiceEnd(hlookup);
+
+    if(error_code != NO_ERROR) {
```
Comment (vtjnash) [inline src/win/util.c:1756]: style nit
```suggestion
    if (error_code != NO_ERROR) {
```

File: src/win/util.c (hunk 5 above)
Comment (kampeador) [reply src/win/util.c:1683]: What compiler are you using?
-std=c89 -Wall -Wextra -pedantic compiles fine with MinGW.
How is this different from this one?
https://github.com/libuv/libuv/blob/v1.x/src/win/tcp.c#L1597

Comment (vtjnash) [reply src/win/util.c:1683]: ah, my mistake. I'm using the wetware compiler, but

Comment (vtjnash) [reply src/win/util.c:1683]: constant integer expressions are also allowable (https://en.cppreference.com/w/c/language/array#Arrays_of_constant_known_size)

**Comment by vtjnash** [comment]
CI: https://ci.nodejs.org/job/libuv-test-commit-windows-cmake/399/

**Review by vtjnash — APPROVED** [review]

**Comment by vtjnash** [comment]
Would be good if a second person could look through and review the implementation meets libuv's standards, to the spec described in https://github.com/libuv/libuv/pull/3285#issuecomment-908691202

**Comment by bnoordhuis** [comment]
I glanced at it but this is somewhat outside my area of expertise. I appreciate @kampeador's work but at 150 lines added I don't think it qualifies as an unobtrusive fix.

File: src/win/util.c [hunk 10]
```diff
@@ -1663,6 +1673,126 @@ int uv_os_unsetenv(const char* name) {
...
+      error_code = WSAGetLastError();
+
+      if ((error_code == WSAEFAULT) && (dwlen > WSAQ_LOCAL_BUF_LEN)) {
+        /* Should never happen */
+        assert(sizeof(CHAR) * dwlen >= sizeof(WSAQUERYSETW));
+
+        /* Fallback to the heap allocation */
+        heap_data = uv__malloc(sizeof(CHAR) * (size_t) dwlen);
+        if (heap_data != NULL)
+        {
```
Comment (zcbenz) [inline src/win/util.c:1741]:
```suggestion
        if (heap_data != NULL) {
```

File: src/win/util.c [hunk 11]
```diff
@@ -1663,6 +1673,126 @@ int uv_os_unsetenv(const char* name) {
...
+  error_code = WSALookupServiceBeginW(pwsaq, LUP_RETURN_NAME, &hlookup);
+  if (error_code == NO_ERROR) {
+    /* Try stack allocation first */
+    error_code = WSALookupServiceNextW(hlookup, 0, &dwlen, pwsaq);
+    if (error_code == NO_ERROR) {
+      result_name = pwsaq->lpszServiceInstanceName;
+    } else {
+      error_code = WSAGetLastError();
+
+      if ((error_code == WSAEFAULT) && (dwlen > WSAQ_LOCAL_BUF_LEN)) {
```
Comment (zcbenz) [inline src/win/util.c:1734]: It seems that there is no need to test `dwlen > WSAQ_LOCAL_BUF_LEN` if result is `WSAEFAULT`?

**Review by zcbenz — COMMENTED** [review]
(I'm not a libuv maintainer, just someone who wants to get this merged 😉 )

**Comment by Biswa96** [comment]
I do not mean to hurt author's attempt. The code seems to be a reverse engineered version of what ws2_32 does internally. Should a project depend on internal behavior of a closed source operating system?

**Comment by vtjnash** [comment]
wepoll is quite similar in that respect. It is the purpose of libuv to abstract away the internal behaviors of various OS.

File: src/win/util.c (hunk 11 above)
Comment (kampeador) [reply src/win/util.c]: The reason I've added it in the first place was that it looked strange to me that WSAEFAULT were used as buffer size check and there can be other cases or hidden bugs. It turned out they've used WSAENIVAL instead.

**Review by zcbenz — APPROVED** [review]

**Comment by zcbenz** [comment]
Bumping to remove the stale label.

**Comment by gresolio** [comment]
Hi everybody,

in the issue #3260, I read that libuv needs additional reviewers to accept this patch.
I'm willing to help, but I don’t know yet how I can be useful.
There is experience with C (mostly embedded) and a little bit with Win32 API.

I have quickly tested the uv__gethostnamew_nt60 function separately from libuv, so far there are no obvious issues:
https://gist.github.com/gresolio/cc149a2ae2dfabefd69d483e2f7ac182

Please tell me what I should pay attention to, what is desirable to evaluate?

**Comment by XVilka** [comment]
Are there any updates on this PR? It would be awesome to have Windows 7 support restored in the next libuv release - it stops us from updating it.

**Comment by zbyna** [comment]
Any update or plan for merging this?

**Comment by zbyna** [comment]
Just close it if there is no intent to merge.

**Comment by XVilka** [comment]
Any decision on this yet?

**Comment by zbyna** [comment]
Any updates?

File: src/win/winapi.h [hunk 12]
```diff
@@ -4739,6 +4739,11 @@ typedef struct _TCP_INITIAL_RTO_PARAMETERS {
 # define  SIO_TCP_INITIAL_RTO _WSAIOW(IOC_VENDOR,17)
 #endif
 
+/* From winsock2.h */
```
Comment (miniak) [inline src/win/winapi.h:4742]: this change should be reverted

File: src/win/winapi.c [hunk 13]
```diff
@@ -46,7 +46,7 @@ sPowerRegisterSuspendResumeNotification pPowerRegisterSuspendResumeNotification;
 sSetWinEventHook pSetWinEventHook;
 
 /* ws2_32.dll function pointer */
-uv_sGetHostNameW pGetHostNameW;
```
Comment (miniak) [inline src/win/winapi.c:49]: this change should be reverted

File: src/win/util.c [hunk 14]
```diff
@@ -1674,10 +1803,10 @@ int uv_os_gethostname(char* buffer, size_t* size) {
 
   uv__once_init(); /* Initialize winsock */
 
-  if (pGetHostNameW == NULL)
-    return UV_ENOSYS;
+  sGetHostNameW gethostnamew =
```
Comment (miniak) [inline src/win/util.c:1806]:
```cpp
  uv_sGetHostNameW gethostnamew =
      pGetHostNameW == NULL ? uv__gethostnamew_nt60 : pGetHostNameW;
```

**Review by miniak — CHANGES_REQUESTED** [review]

**Comment by FlyingDR** [comment]
@kampeador It looks like we're very close to getting this merged, could you please consider addressing the latest review comments? Thank you in advance.

I hope it will be enough for project maintainers.

**Comment by XVilka** [comment]
@FlyingDR more updated version of this PR: https://github.com/libuv/libuv/pull/3636
I don't have time to continue it though (we removed dependency on libuv because of that), feel free to overtake it.

**Comment by vtjnash** [comment]
Moved to https://github.com/libuv/libuv/pull/3636

**Pull request closed without merging**